csharp/ql/test/library-tests/dataflow/arrays/ArrayFlow.cs

```diff
@@ -0,0 +1,229 @@
+        Sink(DictFirstValueB(dict)); // flow [MISSING]
+        Sink(DictFirstValueC(dict)); // flow
+    }
+
+    public void M14(A other)
+    {
+        var dict = new Dictionary<int, A>() { { 0, other } };
+        Sink(dict[0]); // no flow
+        SinkDictValue(dict); // no flow
+        Sink(DictFirstValueA(dict)); // no flow
+        Sink(DictFirstValueB(dict)); // no flow
+        Sink(DictFirstValueC(dict)); // no flow
+    }
+
+    public void M15()
+    {
+        var a = new A();
+        var dict = new Dictionary<A, int>() { { a, 0 } };
+        Sink(dict.Keys.First()); // flow [MISSING]
+        SinkDictKey(dict); // flow [MISSING]
+        Sink(DictFirstKeyA(dict)); // flow [MISSING]
+        Sink(DictFirstKeyB(dict)); // flow [MISSING]
+    }
+
+    public void M16(A other)
+    {
+        var dict = new Dictionary<A, int>() { { other, 0 } };
+        Sink(dict.Keys.First()); // no flow
+        SinkDictKey(dict); // no flow
+        Sink(DictFirstKeyA(dict)); // no flow
+        Sink(DictFirstKeyB(dict)); // no flow
+    }
+
+    public void M17()
+    {
+        var a = new A();
+        var @as = new[] { a };
+        foreach (var x in @as)
+            Sink(x); // flow
+    }
+
+    public void M18(A other)
+    {
+        var @as = new[] { other };
+        foreach (var x in @as)
+            Sink(x); // no flow
+    }
+
+    public void M19()
+    {
+        var a = new A();
+        var @as = new[] { a };
+        var enumerator = @as.GetEnumerator();
+        while (enumerator.MoveNext())
+            Sink(enumerator.Current); // flow
+    }
+
+    public void M20(A other)
+    {
+        var @as = new[] { other };
+        var enumerator = @as.GetEnumerator();
+        while (enumerator.MoveNext())
+            Sink(enumerator.Current); // no flow
+    }
+
+    public void M21()
+    {
+        var a = new A();
+        var list = new List<A>();
+        list.Add(a);
+        var enumerator = list.GetEnumerator();
+        while (enumerator.MoveNext())
+            Sink(enumerator.Current); // flow [MISSING]
+    }
+
+    public static void Sink<T>(T t) { }
+
+    public static void SinkElem<T>(T[] ts) => Sink(ts[0]);
+
+    public static void SinkListElem<T>(IList<T> list) => Sink(list[0]);
+
+    public static void SinkDictValue<T>(IDictionary<int, T> dict) => Sink(dict[0]);
+
+    public static void SinkDictKey<T>(IDictionary<T, int> dict) => Sink(dict.Keys.First());
+
+    public static T First<T>(T[] ts) => ts[0];
+
+    public static T ListFirst<T>(IList<T> list) => list[0];
+
+    public static T DictFirstValueA<T>(IDictionary<int, T> dict) => dict[0];
+
+    public static T DictFirstValueB<T>(IDictionary<int, T> dict) => dict.First().Value;
+
+    public static T DictFirstValueC<T>(IDictionary<int, T> dict) => dict.Values.First();
+
+    public static T DictFirstKeyA<T>(IDictionary<T, int> dict) => dict.Keys.First();
+
+    public static T DictFirstKeyB<T>(IDictionary<T, int> dict) => dict.First().Key;
```

Comment:
Would like slightly more descriptive names than -A, -B, -C suffixes. How about
`DictIndex0`, `DictFirstValue`, `DictValuesFirst`, `DictKeysFirst` and `DictFirstKey`?